lib/evse_security/evse_security.cpp

```diff
@@ -714,34 +715,79 @@ void EvseSecurity::update_ocsp_cache(const CertificateHashData& certificate_hash
                                      const std::string& ocsp_response) {
     std::lock_guard<std::mutex> guard(EvseSecurity::security_mutex);
 
+    // TODO(ioan): shouldn't we also do this for the MO?
     const auto ca_bundle_path = this->ca_bundle_path_map.at(CaCertificateType::V2G);
 
     try {
         X509CertificateBundle ca_bundle(ca_bundle_path, EncodingFormat::PEM);
-        const auto certificates_of_bundle = ca_bundle.split();
+        auto& certificate_hierarchy = ca_bundle.get_certficate_hierarchy();
 
-        for (const auto& cert : certificates_of_bundle) {
-            if (cert == certificate_hash_data) {
-                EVLOG_debug << "Writing OCSP Response to filesystem";
-                if (!cert.get_file().has_value()) {
-                    continue;
-                }
+        try {
+            // Find the certificate
+            X509Wrapper cert = certificate_hierarchy.find_certificate(certificate_hash_data);
+
+            EVLOG_debug << "Writing OCSP Response to filesystem";
+            if (cert.get_file().has_value()) {
                 const auto ocsp_path = cert.get_file().value().parent_path() / "ocsp";
                 if (!fs::exists(ocsp_path)) {
                     fs::create_directories(ocsp_path);
                 }
+
                 const auto ocsp_file_path =
```

Comment:
Since we can now have sub cas stored in the same file, defining the `ocsp_file_path` this way is not a good idea, because it might be overriden everytime if multiple sub cas are stored in the same chain file, which is true for the V2G_CERT_CHAIN.

Reply:
I do not think I completely understand. This only takes into account the root, if the file is over-written, then yes the OCSP data becomes obsolete because it is not linked to the proper root any more.

Reply:
The relevant certificate chain where this function is used for is the SECC leaf chain. The subcas are stored in a chain file and the leaf is stored seperately. This function is called for all certificates of the chain. Since these certificates are stored in the same chain file, the ocsp response would be overriden with every execution for certificates that are actually different. 

Using random filenames and using another approach to search for the correct OCSP response in retrieve_ocsp should solve this though